pkg/ruler/ruler.go

```diff
@@ -1146,24 +1155,47 @@ func (r *Ruler) getLocalRules(userID string, rulesRequest RulesRequest, includeB
 		}
 	}
 
-	if !includeBackups {
-		return groupDescs, nil
+	combinedRuleStateDescs := groupDescs
+	if includeBackups {
+		backupGroups := r.manager.GetBackupRules(userID)
+		backupGroupDescs, err := r.ruleGroupListToGroupStateDesc(userID, backupGroups, groupListFilter{
+			ruleNameSet,
+			ruleGroupNameSet,
+			fileSet,
+			returnAlerts,
+			returnRecording,
+			matcherSets,
+		})
+		if err != nil {
+			return RulesResponse{}, err
+		}
+		combinedRuleStateDescs = append(combinedRuleStateDescs, backupGroupDescs...)
 	}
 
-	backupGroups := r.manager.GetBackupRules(userID)
-	backupGroupDescs, err := r.ruleGroupListToGroupStateDesc(userID, backupGroups, groupListFilter{
-		ruleNameSet,
-		ruleGroupNameSet,
-		fileSet,
-		returnAlerts,
-		returnRecording,
-		matcherSets,
-	})
-	if err != nil {
-		return nil, err
+	if rulesRequest.MaxRuleGroups <= 0 {
+		return RulesResponse{
+			Groups:    combinedRuleStateDescs,
+			NextToken: "",
+		}, nil
 	}
 
-	return append(groupDescs, backupGroupDescs...), nil
+	sort.Sort(GroupStateDescs(combinedRuleStateDescs))
+
+	resultingGroupDescs := make([]*GroupStateDesc, 0, len(combinedRuleStateDescs))
+	for _, group := range combinedRuleStateDescs {
+		groupID := GetRuleGroupNextToken(group.Group.Namespace, group.Group.Name)
+		if len(rulesRequest.NextToken) > 0 && rulesRequest.NextToken >= groupID {
```

Comment:
there is no meaning in comparing two hashes sha1() using `>=`. We should use `==` only.

Reply:
discussed offline, this is ok because are assuming that `combinedRulesStateDescs` is sorted by sha.